repository/state/state.go

```diff
@@ -141,37 +129,256 @@ func (st *State) getOrCreateIdForChecksum(checksum objects.Checksum) uint64 {
-	if len(serialized) < 4 {
-		return nil, fmt.Errorf("invalid state data")
+func (st *State) SerializeStream(w io.Writer) error {
+	writeUint64 := func(value uint64) error {
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, value)
+		_, err := w.Write(buf)
+		return err
+	}
+
+	writeUint32 := func(value uint32) error {
+		buf := make([]byte, 4)
+		binary.LittleEndian.PutUint32(buf, value)
+		_, err := w.Write(buf)
+		return err
+	}
+
+	writeLocation := func(loc Location) error {
+		if err := writeUint64(loc.Packfile); err != nil {
+			return err
+		}
+		if err := writeUint32(loc.Offset); err != nil {
+			return err
+		}
+		return writeUint32(loc.Length)
+	}
+
+	// Serialize Metadata
+	if err := writeUint32(st.Metadata.Version); err != nil {
+		return fmt.Errorf("failed to write version: %w", err)
+	}
+	timestamp := st.Metadata.Timestamp.UnixNano()
+	if err := writeUint64(uint64(timestamp)); err != nil {
+		return fmt.Errorf("failed to write timestamp: %w", err)
+	}
+	if st.Metadata.Aggregate {
+		if _, err := w.Write([]byte{1}); err != nil {
+			return fmt.Errorf("failed to write aggregate flag: %w", err)
+		}
+	} else {
+		if _, err := w.Write([]byte{0}); err != nil {
+			return fmt.Errorf("failed to write aggregate flag: %w", err)
+		}
+	}
+	if err := writeUint64(uint64(len(st.Metadata.Extends))); err != nil {
+		return fmt.Errorf("failed to write extends length: %w", err)
+	}
+	for _, checksum := range st.Metadata.Extends {
+		if _, err := w.Write(checksum[:]); err != nil {
+			return fmt.Errorf("failed to write checksum: %w", err)
+		}
+	}
+
+	if err := serializeMapping(w, st.DeletedSnapshots, func(key uint64) error {
+		return writeUint64(key)
+	}, func(value time.Time) error {
+		return writeUint64(uint64(value.UnixNano()))
+	}); err != nil {
+		return fmt.Errorf("failed to serialize DeletedSnapshots: %w", err)
 	}
 
-	serialized, versionBytes := serialized[:len(serialized)-4], serialized[len(serialized)-4:]
-	version := binary.LittleEndian.Uint32(versionBytes)
-	if version != VERSION {
-		return nil, fmt.Errorf("invalid state version: %d", version)
+	// Serialize each mapping
+	if err := serializeMapping(w, st.IdToChecksum, writeUint64, func(v objects.Checksum) error { _, err := w.Write(v[:]); return err }); err != nil {
+		return fmt.Errorf("failed to serialize IdToChecksum: %w", err)
 	}
 
-	var st State
-	if err := msgpack.Unmarshal(serialized, &st); err != nil {
-		return nil, err
+	mappings := []struct {
+		name string
+		data map[uint64]Location
+	}{
+		{"Chunks", st.Chunks},
+		{"Objects", st.Objects},
+		{"Files", st.Files},
+		{"Directories", st.Directories},
+		{"Children", st.Children},
+		{"Datas", st.Datas},
+		{"Snapshots", st.Snapshots},
+		{"Signatures", st.Signatures},
+		{"Errors", st.Errors},
 	}
 
-	st.rebuildChecksums()
+	for _, m := range mappings {
+		if err := serializeMapping(w, m.data, writeUint64, func(v Location) error {
+			return writeLocation(v)
+		}); err != nil {
```

Comment:
nit: could pass `writeLocation` directly instead of wrapping it in a no-op.

Reply:
fixed !